web/src/main/java/org/cbioportal/web/util/AlterationCountUtil.java

```diff
@@ -0,0 +1,109 @@
+            Select.all(): Select.byValues(entrezGeneIds);
+        List<MolecularProfileCaseIdentifier> molecularProfileCaseIdentifiers =
+            alterationCountFilter.getMolecularProfileCaseIdentifiers();
+        boolean includeFrequency = true;
+        boolean includeMissingAlterationsFromGenePanel = false;
+        AlterationFilter alterationFilter = alterationCountFilter.getAlterationFilter();
+
+         return alterationCountType.equals(AlterationCountType.SAMPLE) ?
+            alterationCountService.getSampleAlterationCounts(
+                molecularProfileCaseIdentifiers,
+                geneSelect,
+                includeFrequency,
+                includeMissingAlterationsFromGenePanel,
+                alterationFilter
+            ) :
+            alterationCountService.getPatientAlterationCounts(
+                molecularProfileCaseIdentifiers,
+                geneSelect,
+                includeFrequency,
+                includeMissingAlterationsFromGenePanel,
+                alterationFilter
+            );
+    }
+    
+    public Set<GenePanelToGene> getProfiledGenes(
+        AlterationCountFilter alterationCountFilter,
+        AlterationCountType alterationCountType,
+        List<Integer> entrezGeneIds
+    ) {
+        List<MolecularProfileCaseIdentifier> molecularProfileCaseIdentifiers =
+            alterationCountFilter.getMolecularProfileCaseIdentifiers();
+        Set<GenePanelToGene> profiledGenes;
+        List<GenePanelData> genePanelDataList = alterationCountType.equals(AlterationCountType.SAMPLE) ?
+            genePanelService.fetchGenePanelDataInMultipleMolecularProfiles(molecularProfileCaseIdentifiers) :
+            genePanelService.fetchGenePanelDataInMultipleMolecularProfilesByPatientIds(molecularProfileCaseIdentifiers);
+
+        if (genePanelDataList
+            .stream()
+            .anyMatch(
+                genePanelData -> genePanelData.getProfiled() && genePanelData.getGenePanelId() == null
+            )
+        ) {
+            // if there is at least one sample/patient with all genes profiles,
+            // then all genes are profiled for the current data set
+            profiledGenes = null;
+        } else {
+            List<String> profiledPanelIds =  genePanelDataList
+                .stream()
+                .filter(GenePanelData::getProfiled)
+                .map(GenePanelData::getGenePanelId)
+                .collect(Collectors.toList());
+
+            if (profiledPanelIds.isEmpty()) {
+                profiledGenes = Collections.emptySet();
+            } else {
+                List<GenePanel> panels = genePanelService.fetchGenePanels(
+                    profiledPanelIds,
+                    Projection.DETAILED.name()
+                );
+
+                Stream<GenePanelToGene> genePanelGenes = panels
+                    .stream()
+                    .map(GenePanel::getGenes)
+                    .filter(Objects::nonNull)
```

Comment:
The gene panel API is breaking me yet again. Why can this be null? What use is a gene panel that does not refer to a list of genes? Can we extend the gene panel service API to include some methods with better guarantees about gene panels, because this is super counter intuitive

Reply:
I agree. We need some refactoring effort for these kind of things. But I think this is out of this PR's scope.